### PR TITLE
Defensive & Offensive Programming 

### PR DESCRIPTION
**Pull request by tyhoff**
I do want to change the slug from offensive-programming because, well, there are two definitions for that word.

### CODE REVIEW

File: _drafts/offensive-programming.md [hunk 1]
```diff
@@ -0,0 +1,562 @@
...
+If a developer comes along and tries to use an API and passes invalid arguments
+into the function, make sure the application yells at them to fix the issue.
+There is nothing worse than receiving a `-1` return value and digging through 10
+layers of firmware code only to find out the real reason was that you passed in
+a string of 20 characters when the maximum was 16 or because you accidentally
+passed a null pointer due to an uninitialized variable.
+
+```c
+void device_set_name(char *name, size_t name_len) {
+  ASSERT(name && name_len =< 16);
```
Comment (chrisc11) [inline _drafts/offensive-programming.md:216]:
```suggestion
  ASSERT(name && name_len <= 16);
```

File: _drafts/offensive-programming.md [hunk 2]
```diff
@@ -0,0 +1,562 @@
...
+  ...
+}
+```
+
+The only time I would choose, or at least heavily debate, to not use asserts to
+validate arguments is when I am building a library that will be used by people
+outside of my organization. In those cases, I would make the validation asserts
+optional, just as FreeRTOS does with their RTOS functions by allowing developers
+to define ``configASSERT` themselves[^freertos_asserts].
```
Comment (chrisc11) [inline _drafts/offensive-programming.md:225]:
```suggestion
to define `configASSERT` themselves[^freertos_asserts].
```

File: _drafts/offensive-programming.md [hunk 3]
```diff
@@ -0,0 +1,562 @@
...
+```c
+void *malloc_assert(size_t n) {
+  void *p = malloc(n)
+  ASSERT(p);
+  return p;
+  ...
+}
+```
+
+In code where the calls to `malloc()` should never fail, such as RTOS structure
```
Comment (chrisc11) [inline _drafts/offensive-programming.md:254]:
```suggestion
In code where the calls to `malloc()` should never fail, such as for the allocation of RTOS primitives
```

File: _drafts/offensive-programming.md [hunk 4]
```diff
@@ -0,0 +1,562 @@
...
+allocation, request and response buffers, etc., we can use this asserted
+version.
+
+Another common issue with RTOS-based systems is a queue becoming full and
+possibly not emptying quickly enough. As with memory depletion, this isn't a
+show-stopping issue, because the system might recover, but we want to look into
```
Comment (chrisc11) [inline _drafts/offensive-programming.md:260]: Could call out this means we are dropping events which could be important/critical to operation

File: _drafts/offensive-programming.md [hunk 5]
```diff
@@ -0,0 +1,562 @@
...
+
+### Software Stalls & Deadlocks
+
+Embedded devices need to respond in a reasonable amount of time to user input
+and communication packets all while being performant and not showing any signs
+of lag or stalls.
+
+I can't count the number of times while working on previous projects that slow
+flash operations caused the system to freeze for 2-3 seconds at a time, wreaking
+havoc on the user experience causing other operations in the system to time out.
```
Comment (chrisc11) [inline _drafts/offensive-programming.md:308]:
```suggestion
havoc on the user experience or causing other operations in the system to time out.
```

File: _drafts/offensive-programming.md [hunk 6]
```diff
@@ -0,0 +1,562 @@
...
+The worst part about these issues is that they often aren't brought to our
+attention until the end-users complain.
+
+To help catch these issues before pushing firmware to external users, you can
+create and configure your [task
```
Comment (chrisc11) [inline _drafts/offensive-programming.md:313]: doesn't render as a link

File: _drafts/offensive-programming.md [hunk 7]
```diff
@@ -0,0 +1,562 @@
...
+system calls.
+
+```c
+void timing_sensitive_task(void) {
+  const bool success = mutex_lock(&s_mutex, 1000 /* wait 1s */);
+  ASSERT(success);
+  {
+    ...
+  }
+}
```
Comment (chrisc11) [inline _drafts/offensive-programming.md:326]: if we have a task watchdog, do we need the explicit timeout?

File: _drafts/offensive-programming.md [hunk 8]
```diff
@@ -0,0 +1,562 @@
...
+Thankfully, those strings don't get included in the final firmware binary so you
+are free to make them as long as you like!
+
+## Handling External & Application Code
+
+There are places where one would not want to use offensive programming
+practices. Primarily, it's when the developer is not fully in control of the
+software, hardware, or incoming data. These could be any of the following:
+
+- Hardware & hardware drivers
```
Comment (chrisc11) [inline _drafts/offensive-programming.md:428]:
```suggestion
- Hardware & peripheral drivers
```

File: _drafts/offensive-programming.md [hunk 9]
```diff
@@ -0,0 +1,562 @@
...
+record issues, we'll never know what truly causes these issues and how often
+they occur.
+
+The parts of OffP that should be kept compiled in and enabled in production
+builds are the ones that ensure that undefined behavior is not experienced in
+firmware. This includes argument validation, watchdog timeouts to detect
+deadlocks, bugs potentially leading to memory corruption, etc. It's better to
+return the device to a known state than to limp along.
+
+What we can turn down to level 3/10 or disable completely are the UX-like bugs,
```
Comment (chrisc11) [inline _drafts/offensive-programming.md:491]:
```suggestion
What we can often consider disabling completely are marginal issues, such as software stalls, 
```

File: _drafts/offensive-programming.md [hunk 10]
```diff
@@ -0,0 +1,562 @@
...
+such as software stalls, malloc and queue failures (which are handled), and
+timeouts. These types of issues aren't guaranteed to recover or go away, but
+they may be acceptable depending on the context and how the device is used.
+
+If you choose to reduce the aggressiveness of the checks, consider keeping the
+hooks and timeouts and instead log them somewhere, rather than reset the system.
+For inspiration, check out
+[ARM MBed's Error Handling API](https://os.mbed.com/docs/mbed-os/v6.5/apis/error-handling.html),
+which keeps a circular buffer of errors that the system has experienced in a
```
Comment (chrisc11) [inline _drafts/offensive-programming.md:500]:
```suggestion
which keeps a circular buffer of warnings that the system has experienced in a
```

File: _drafts/offensive-programming.md (hunk 7 above)
Comment (tyhoff) [reply _drafts/offensive-programming.md:326]: Good point. Added something.

File: _drafts/offensive-programming.md (hunk 4 above)
Comment (tyhoff) [reply _drafts/offensive-programming.md:265]: Great point

File: _drafts/offensive-programming.md [hunk 11]
```diff
@@ -0,0 +1,562 @@
...
+  function arguments, and system state such as the heap, lists, and queues. With
+  all this information, developers can root cause issues more quickly.
+- **Faster development** - Writing a new piece of code that integrates into
+  multiple modules can be tricky if there are bugs or assumptions in the modules
+  you are integrating with. By having assertions and OffP practices in the
+  surrounding layers, if a developer does anything misaligned with the API
+  specs, the system will immediately alert the developer. It's better than
+  tracking down what `-3` or `known_error` means and how it was bubbled up
+  through the system.
+- **Awareness of prevalence** - No developer or customer notices bugs unless the
```
[the remark this answers is not in the document]
Comment (tyhoff) [reply _drafts/offensive-programming.md]: Done!

File: _drafts/offensive-programming.md [hunk 12]
```diff
@@ -0,0 +1,560 @@
...
+track down those 1 in 1,000 hour bugs, effeciently root cause them, and keep
+your end-users happy.
+
+{% include newsletter.html %}
+
+{% include toc.html %}
+
+## Defensive Programming
+
+Defensive programming (referred to from here on as DefP) is the practice of
```
[the remark this answers is not in the document]
Comment (tyhoff) [reply _drafts/offensive-programming.md]: Removed abbreviations. Thanks for the push.

File: _drafts/offensive-programming.md [hunk 13]
```diff
@@ -0,0 +1,599 @@
+---
+title: Defensive Programming - Friend or Foe?
+description:
+  Embedded systems can benefit from defensive and offensive programming with the
+  use of asserts, timeouts, and watchdogs.
+author: tyler
+image: img/offensive-programming/cover.png
+---
+
+One of the worst things about embedded development (especially in C) is
```
Comment (franc0is) [inline _drafts/offensive-programming.md]: FWIW I think the slightly shorter introduction you had before was better (but not a big deal).

**Review by chrisc11 — APPROVED** [review]

File: _drafts/offensive-programming.md [hunk 14]
```diff
@@ -0,0 +1,599 @@
...
+they occur.
+
+The parts of OffP that should be kept compiled in and enabled in production
+builds are the ones that ensure that undefined behavior is not experienced in
+firmware. This includes argument validation, watchdog timeouts to detect
+deadlocks, bugs potentially leading to memory corruption, etc. It's better to
+return the device to a known state than to limp along.
+
+What we can often consider disabling completely are marginal issues, such as
+software stalls, such as software stalls, malloc and queue failures (which are
```
Comment (franc0is) [inline _drafts/offensive-programming.md]:
```suggestion
software stalls, malloc and queue failures (which are
```

**Review by franc0is — APPROVED** [review]
Excellent post, I found it engaging and interesting. Ship it!

**Pull request merged**